pkg/test/server.go

```diff
@@ -0,0 +1,125 @@
+package test
+
+import (
+	"encoding/csv"
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/miekg/dns"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+var domainToIP = loadRecords("testdata/test-300.csv")
+
+const truncateSize = 512
+
+type handler struct {
+	truncate bool
+}
+
+type Server struct {
+	Addr    string
+	TCPPort string
+	UDPPort string
+}
+
+func (s *Server) Run() {
+	g := new(errgroup.Group)
+	udpSrv := &dns.Server{Addr: net.JoinHostPort(s.Addr, s.UDPPort), Net: "udp", Handler: &handler{truncate: true}}
+	defer func() {
+		if err := udpSrv.Shutdown(); err != nil {
+			logrus.Errorf("shutting down UDP server: %v", err)
+		}
+	}()
+	g.Go(udpSrv.ListenAndServe)
+
+	tcpSrv := &dns.Server{Addr: net.JoinHostPort(s.Addr, s.TCPPort), Net: "tcp", Handler: &handler{}}
+	defer func() {
+		if err := tcpSrv.Shutdown(); err != nil {
+			logrus.Errorf("shutting down TCP server: %v", err)
+		}
+	}()
+	g.Go(tcpSrv.ListenAndServe)
+
+	if err := g.Wait(); err != nil {
+		logrus.Fatalf("ListenAndServe error: %v", err)
+	}
+}
+
+func (h *handler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	msg := new(dns.Msg)
+	msg.SetReply(r)
+	msg.Compress = false
+	// we might handle other OpCodes later
+	if r.Opcode != dns.OpcodeQuery {
+		return
+	}
+	parseReply(msg)
+	if h.truncate {
+		msg.Truncate(truncateSize)
+	}
+	err := w.WriteMsg(msg)
+	if err != nil {
+		logrus.Errorf("ServeDNS trying to write message in udpHandler: %v", err)
+	}
+}
+
+func parseReply(msg *dns.Msg) {
+	logrus.Infof("ServeDNS handling request: %v", msg.Question)
+	for _, q := range msg.Question {
+		switch q.Qtype { //nolint:gocritic // this will have additional cases soon
+		case dns.TypeA:
+			msg.Authoritative = true
```

Comment:
Do we expect to be authoritative? I wouldn't expect it to be for the general case (I'd expect the upstream of user systems to be the DNS server at their ISP, which would typically not be authoritative for most of the internet.)

Reply:
This will be the final upstream server and will not be acting as a recursive server since the plan is not to have any lookup using the net. Also, win32 API will not work without the internet.